### PR TITLE
Allowing empty fields for mutations

### PR DESCRIPTION
**Pull request by ohager**
### Requirements

* Filling out the template is required. Any pull request that does not include enough information to be reviewed in a timely manner may be closed at the maintainers' discretion.
* All new code requires tests to ensure against regressions

### Description of the Change

Empty fields arrays create a syntactically incorrect GraphQL expression, which is totally valid for 'query' operations, but not for 'mutation'. This PR checks for empty arrays of a query and eventually throws an error (leading to a slightly more consistent behavior), and furthermore allows mutations with empty/void return values. Although, the latter is not considered as best practice, it makes gotql usable for situations where such best practice is not applied.

(see #39)

### Alternate Designs

Alternatively, a specific `void` flag was proposed, but I thought the current solution has a bit less overhead in terms of implement, as no additional parameter needs to be introduced. From the perspective of developer experience, the _implicit_ option is much more intuitive and has less cognitive overhead.

### Benefits

This PR makes this very useful (and well written) lib even more robust.

### Possible Drawbacks

Allowing mutations without return values breaks with nowadays best practices (that's why I added an informative hint about godd practices)

### Applicable Issues

see #39

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -149,3 +149,6 @@ $RECYCLE.BIN/
 *.lnk
 
 # End of https://www.gitignore.io/api/node,macos,visualstudiocode,vim,windows,linux
+
+### Jetbrains ###
+.idea
```
Comment (ohager) [inline .gitignore:154]: For those who use webstorm :grin:

File: src/modules/parser.ts [hunk 2]
```diff
@@ -221,14 +221,18 @@ function checkArgs (argsList: QueryType['operation']['args'], operationArg: stri
...
-function parseOperation (query: QueryType): string {
+function parseOperation (query: QueryType, allowEmptyFields: boolean): string {
   let operation = query.operation
+  const hasEmptyFields = !operation.fields || !operation.fields.length
   info('Parsing operation %s', operation.name)
   if (!operation.name) throw new Error(`name is required for graphQL operation`)
-  if (!operation.fields) throw new Error(`field list is required for operation "${operation.name}"`)
+  if (hasEmptyFields && !allowEmptyFields) throw new Error(`field list is required for operation "${operation.name}"`)
+
+  if(hasEmptyFields) info('Hint: having no fields is not considered as a good practice')
```
Comment (ohager) [inline src/modules/parser.ts:235]: IMHO this hint is useful for unexperienced GraphQL devs, like me

File: tests/parser.test.js [hunk 3]
```diff
@@ -541,11 +541,11 @@ describe('Should allow for nested arguments in mutation or query operations (#28
         id: '1234',
         data: {
           name: 'foo',
-          status: 'bar',
-        },
+          status: 'bar'
```
Comment (ohager) [inline tests/parser.test.js:544]: I ran `lint fix` here...

File: src/modules/parser.ts (hunk 2 above)
Comment (khaosdoctor) [reply src/modules/parser.ts:235]: I agree!

File: src/modules/parser.ts [hunk 4]
```diff
@@ -221,14 +221,18 @@ function checkArgs (argsList: QueryType['operation']['args'], operationArg: stri
 
 /**
  * Parses the operation bit of the query
- * @param {queryType} query The JSON-Like query to be parsed
+ * @param {QueryType} query The JSON-Like query to be parsed
+ * @param {boolean} allowEmptyFields If 'true', empty fields are allowed (which is not considered a good practice)
  * @return {string} Parsed operation query
  */
-function parseOperation (query: QueryType): string {
+function parseOperation (query: QueryType, allowEmptyFields: boolean): string {
```
Comment (khaosdoctor) [inline src/modules/parser.ts:228]: The problem here is that you add a new **required** parameter. This introduces a breaking change. So my suggestion would be this:

```suggestion
function parseOperation (query: QueryType, allowEmptyFields: boolean = false): string {
```

File: src/modules/parser.ts [hunk 5]
```diff
@@ -221,14 +221,18 @@ function checkArgs (argsList: QueryType['operation']['args'], operationArg: stri
 
 /**
  * Parses the operation bit of the query
- * @param {queryType} query The JSON-Like query to be parsed
+ * @param {QueryType} query The JSON-Like query to be parsed
+ * @param {boolean} allowEmptyFields If 'true', empty fields are allowed (which is not considered a good practice)
```
Comment (khaosdoctor) [inline src/modules/parser.ts:225]:
```suggestion
 * @param {boolean} allowEmptyFields If 'true', empty fields are allowed (WARNING: this is not considered a good practice)
```

File: src/modules/parser.ts [hunk 6]
```diff
@@ -260,9 +266,10 @@ export function parse (query: QueryType, type: GotQL.ExecutionType): string {
     if (!query.operation) throw new Error('a query must have at least one operation')
     if (!type) throw new Error('type must be either "query" or "mutation"')
 
+    let isMutation = type === 'mutation'
```
Comment (khaosdoctor) [inline src/modules/parser.ts:269]: This is not needed. There's a type called `GotQL.ExecutionType` which is an `enum` with two values: MUTATION and QUERY.

```suggestion

```

File: src/modules/parser.ts [hunk 7]
```diff
@@ -260,9 +266,10 @@ export function parse (query: QueryType, type: GotQL.ExecutionType): string {
...
     let queryName = (query.name) ? `${query.name} ` : ''
     info('Defining name "%s"', queryName)
-    const parsedQuery = `${type.trim()} ${queryName}${getQueryVars(query.variables)}{ ${parseOperation(query)} }`.trim()
+    const parsedQuery = `${type.trim()} ${queryName}${getQueryVars(query.variables)}{ ${parseOperation(query, isMutation)} }`.trim()
```
Comment (khaosdoctor) [inline src/modules/parser.ts:272]: Use it here

```suggestion
    const parsedQuery = `${type.trim()} ${queryName}${getQueryVars(query.variables)}{ ${parseOperation(query, type === GotQL.ExecutionType.MUTATION)} }`.trim()
```

File: src/modules/parser.ts (hunk 2 above)
Comment (khaosdoctor) [inline src/modules/parser.ts:235]:
```suggestion
  if(hasEmptyFields) info('Hint: Returning no fields is not considered a good practice')
```

File: src/modules/parser.ts (hunk 4 above)
Comment (ohager) [reply src/modules/parser.ts]: had the same thought initially, although it is kind of "private" function...so I decided against

File: src/modules/parser.ts (hunk 7 above)
Comment (ohager) [reply src/modules/parser.ts]: good catch

File: src/modules/parser.ts (hunk 6 above)
Comment (ohager) [reply src/modules/parser.ts]: good catch

File: src/modules/parser.ts (hunk 2 above)
Comment (ohager) [reply src/modules/parser.ts]: sounds better, yeah

**Pull request merged**